src/Stratis.Bitcoin/Database/ReadWriteBatch.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Database
+{
+    /// <summary>
+    /// A batch that can be used to record changes that can be applied atomically.
+    /// </summary>
+    /// <remarks>The supplied <see cref="Get"/> method will immediately reflect any changes that have 
+    /// been made or retrieve the value from the underlying database. In contrast the database <see cref="IDb.Get"/> method
+    /// will only show the changes after the <see cref="Write"/> method is called.</remarks>
+    public class ReadWriteBatch : IDbBatch
+    {
+        private readonly IDb db;
+        private readonly IDbBatch batch;
+        private Dictionary<byte[], byte[]> cache;
+
+        public ReadWriteBatch(IDb db, params byte[] tables)
+        {
+            this.db = db;
+            this.batch = db.GetWriteBatch(tables);
+            this.cache = new Dictionary<byte[], byte[]>(new ByteArrayComparer());
+        }
+
+        /// <summary>
+        /// Records a value that will be written to the database when the <see cref="Write"/> method is invoked.
+        /// </summary>
+        /// <param name="key">The table key that identifies the value to be updated.</param>
+        /// <param name="value">The value to be written to the table.</param>
+        /// <returns>This class for fluent operations.</returns>
+        public IDbBatch Put(byte[] key, byte[] value)
+        {
+            this.cache[key] = value;
+            return this.batch.Put(key, value);
+        }
+
+        /// <summary>
+        /// Records a value that will be written to the database when the <see cref="Write"/> method is invoked.
+        /// </summary>
+        /// <param name="table">The table that will be updated.</param>
+        /// <param name="key">The table key that identifies the value to be updated.</param>
+        /// <param name="value">The value to be written to the table.</param>
+        /// <returns>This class for fluent operations.</returns>
```

Comment:
The usage of class vs interface in the `<returns>` comments is a bit inconsistent

Reply:
Fixed.